### PR TITLE
CFE-3256: Added warning if CSV parser parses nothing from non-empty file

### PR DESCRIPTION
**Pull request by karlhto**
https://tracker.mender.io/browse/CFE-3256

### CODE REVIEW

File: libutils/json-utils.c [hunk 1]
```diff
@@ -254,6 +254,15 @@ bool JsonParseCsvFile(const char *input_path, size_t size_max, JsonElement **jso
 
     JsonElement *const json = JsonArrayCreate(50);
 
+    if (feof(fin))
+    {
+        json_out = json;
+        Log(LOG_LEVEL_VERBOSE,
+            "%s: CSV file '%s' was empty, so nothing was parsed"
```
Comment (olehermanse) [inline libutils/json-utils.c]: missing comma after this argument.

File: libutils/json-utils.c [hunk 2]
```diff
@@ -297,6 +306,15 @@ bool JsonParseCsvFile(const char *input_path, size_t size_max, JsonElement **jso
         return false;
     }
 
+    if (JsonLength(json) == 0)
+    {
+        Log(LOG_LEVEL_WARNING,
+            "%s: CSV file '%s' is not empty, but nothing was parsed",
+            myname);
```
Comment (olehermanse) [inline libutils/json-utils.c]: Missing second arg, file name.

File: libutils/json-utils.c [hunk 3]
```diff
@@ -297,6 +306,15 @@ bool JsonParseCsvFile(const char *input_path, size_t size_max, JsonElement **jso
...
+        Log(LOG_LEVEL_WARNING,
+            "Make sure the file contains DOS (CRLF) line endings");
```
Comment (vpodzime) [inline libutils/json-utils.c]: Could we make it this only show up if there are newlines in the file? Something like `fseek(fin, 0); if (contains_newline(fin))...`.

**Comment by karlhto** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4596/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4596/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4596/

File: libutils/json-utils.c [hunk 4]
```diff
@@ -93,6 +93,37 @@ static char *filtered_copy(const char *src, char *dst)
...
+    while (!feof(fp))
+    {
+        if (c == '\n')
+        {
+            res = true;
+            break;
+        }
+
+        c = getc(fp);
+    }
```
Comment (vpodzime) [inline libutils/json-utils.c]: I think this should be reading larger chunks, say 256 bytes at a time.

File: libutils/json-utils.c [hunk 5]
```diff
@@ -286,17 +326,29 @@ bool JsonParseCsvFile(const char *input_path, size_t size_max, JsonElement **jso
...
     }
+    else if (JsonLength(json) == 0)
+    {
+        Log(LOG_LEVEL_WARNING,
+            "%s: CSV file '%s' is not empty, but nothing was parsed",
+            myname, input_path);
+        if (contains_newline(fin))
+        {
+            Log(LOG_LEVEL_WARNING,
+                "Make sure the file contains DOS (CRLF) line endings");
```
Comment (vpodzime) [inline libutils/json-utils.c]: *Make sure the file uses CRLF sequence as record separators (line endings)*

Comment (olehermanse) [reply libutils/json-utils.c]: I think karl's wording is easier to understand.

File: libutils/json-utils.c [hunk 6]
```diff
@@ -286,17 +326,29 @@ bool JsonParseCsvFile(const char *input_path, size_t size_max, JsonElement **jso
...
         JsonDestroy(json);
+        fclose(fin);
         return false;
     }
+    else if (JsonLength(json) == 0)
+    {
+        Log(LOG_LEVEL_WARNING,
+            "%s: CSV file '%s' is not empty, but nothing was parsed",
+            myname, input_path);
+        if (contains_newline(fin))
```
Comment (olehermanse) [inline libutils/json-utils.c]: I don't think this is really necessary. I think the warning should be printed for all non-empty files which don't parse to any data.

Comment (karlhto) [reply libutils/json-utils.c]: Alright, I'll scrap the function then.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4625/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4625/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4625/

**Pull request merged**